virtual_ecosystem/core/core_components.py

```diff
@@ -340,31 +368,153 @@ def __post_init__(self, config: Config) -> None:
                 "layer_roles": ("layers", self.layer_roles),
-                "cell_id": np.arange(self.n_cells),
+                "cell_id": np.arange(self._n_cells),
             },
         )
 
         LOGGER.info("Layer structure built from model configuration")
 
+    def _set_base_index(self, name: str, bool_values: NDArray[np.bool_]) -> None:
+        """Helper method to populate the boolean and integer indices for base roles.
+
+        Args:
+            name: the name of the base role
+            bool_values: the boolean representation of the index data.
+        """
+        index_key = frozenset([name])
+        self._role_indices_bool[index_key] = bool_values
+        self._role_indices_int[index_key] = np.nonzero(bool_values)[0]
+
+    def _set_aggregate_index(self, aggregate_index_key: frozenset) -> None:
+        """Method to validate and set an aggregate role index.
+
+        This method checks an aggregate index key (e.g. ``frozenset(['above','canopy'])
+        is made up of valid base role indices and then populates the underlying role
+        indices dictionaries with the resulting aggregate index across the base roles.
+
+        Args:
+            aggregate_index_key: a role index key containing more than one base role
+                name.
+        """
+
+        # Get list of basic roles as frozensets
+        base_index_keys = [frozenset([r]) for r in aggregate_index_key]
+
+        # Check for unknown roles
+        unknown_roles = [
+            role
+            for role, base_key in zip(aggregate_index_key, base_index_keys)
+            if base_key not in self._role_indices_bool
+        ]
+        if unknown_roles:
+            to_raise = ValueError(f"Unknown layer role(s): {','.join(unknown_roles)}")
+            LOGGER.error(to_raise)
+            raise to_raise
+
+        # Construct aggregate indices for known roles
+        bool_index = np.logical_or.reduce(
+            tuple(self._role_indices_bool[role] for role in base_index_keys)
+        )
+
+        self._role_indices_bool[aggregate_index_key] = bool_index
+        self._role_indices_int[aggregate_index_key] = np.where(bool_index)[0]
+
+    def get_indices(self, roles: str | tuple[str], as_bool: bool = True) -> NDArray:
+        """Get the indices for a role or combination of roles.
+
+        Args:
+            roles: a role name or tuple of role names
+            as_bool: sets whether indices are returned as boolean or integer arrays.
+        """
+
+        # Get the key as a frozen set
+        roles_is_str = isinstance(roles, str)
+        roles_as_tuple = (roles,) if roles_is_str else roles
+        idx_key = frozenset(roles_as_tuple)
+
+        # Get the requested index type
+        if as_bool:
+            idx_dict: dict = self._role_indices_bool
+        else:
+            idx_dict = self._role_indices_int
+
+        # Check if the key is already present
+        if idx_key in idx_dict:
+            return idx_dict[idx_key]
```

Comment:
I comes to my mind that you could do something clever with caching here:

```python
from functools import lru_cache
...

@lru_cache
def get_indices(self, *args: tuple[str], as_bool: bool = True) -> NDArray:
    if len(args) == 0:
        raise ...

    idx_dict = self._role_indices_bool if as_bool else self._role_indices_int
    idx_key = frozenset(args)

    ...
```
What I think is nice about this approach is that:
1. The user does not need to pass explicitly a tuple when more than one role, just writing them one after the other, eg. `get_index("soil", "surface")` (this can also be implemented in your approach).
2. If the function has already been called with exactly the same inputs, the result is cached, so the body is not executed. 
3. You only need to manually store the base indices, as the others would be available via automatic caching - instead of manual caching, which is what you do now within the `_set_aggregate_index`.

Using the builtin caching is up to you, but using `*args` will make your code simpler.

Reply:
That is clever, but we've moved away from the aggregate indices (thanks to your questioning of whether we _really_ needed it 😁 ).